### ISSUE TITLE
Trouble with saving yolov8 customized model

### ISSUE THREAD

**Issue by xizzax** [ultralytics/ultralytics#8110]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

@glenn-jocher hey, i am trying to modify the YOLOv8 classification model to include my own loss function (i am trying to incorporate the idea of feature wise group lasso regularization in loss calculation). i have customized the model according to the Advanced Customization section of the docs, but i can't figure out how to train the model. here is what i have so far: 
```
new_model = GroupSparseClassificationModel(cfg='yolov8n-cls.yaml', nc=100, verbose=True)
print(new_model.state_dict())
torch.save(new_model.state_dict(), 'my_model.pt')
```
```
args = dict(model='my_model.pt', data=dataset, epochs=1)
trainer = GroupSparseClassificationTrainer(overrides=args)
trainer.train()
```
i believe the error lies in saving the new model using `new_model.state_dict()` because it does not save the same values that a pre-built yolo file (like yolov8n-cls.pt) would have. kindly help

**Comment by glenn-jocher** [comment]
@xizzax hello! It looks like you're on the right track with customizing your YOLOv8 classification model. To save your customized model for training, you should save both the model's `state_dict` and the configuration file used to create the model. Here's a concise way to do it:

```python
# Save model weights and configuration
torch.save({'model': new_model.state_dict(), 'yaml': 'yolov8n-cls.yaml'}, 'my_model.pt')

# Load model and train
args = dict(model='my_model.pt', data=dataset, epochs=1)
trainer = GroupSparseClassificationTrainer(overrides=args)
trainer.train()
```

This will ensure that both the weights and the configuration are stored correctly. Give it a try and let us know if you encounter any issues! 👍

**Comment by xizzax** [comment]
I tried this and it gives me the following error: 
![image](https://github.com/ultralytics/ultralytics/assets/76040683/2652009a-883a-4b85-8624-5984110de364)

here is the modified model in case that is relevant:
```
# modifying the model to include group sparsity

class GroupSparseClassificationModel(ClassificationModel):
    def __init__(self, cfg, nc, verbose):
        super().__init__(cfg=cfg, nc=nc, verbose=verbose)
    def init_criterion(self):
        return GroupSparseLoss()

class GroupSparseLoss(v8ClassificationLoss):
    def __call__(self, preds, batch):
        loss, loss_items =  super().__call__(preds, batch)
        loss += self.group_lasso_regularization(0.001)
        loss_items += {'group_lasso': loss}
        return loss, loss_items
    

    def group_lasso_regularization(self, lambda_reg_coeff = 0.1): 
        """implements group sparsity using group lasso regularization"""
        x = 0
        print("REGULARIZING...")
        for n, _module in super().model.named_modules(): # TODO: replace with actual model
            if isinstance(_module, nn.Conv2d) and  (not 'downsample' in n):
                p = _module.weight 
                p = p.reshape(p.shape[0],p.shape[1],p.shape[2]*p.shape[3])
                
                #group lasso regularization based on the feature wise grouping
                x += lambda_reg_coeff*torch.sum(torch.sqrt(torch.sum(torch.sum(p**2,0),1)))

                #group lasso regularization based on the neuron wise grouping
		        #x += _lambda*torch.sum(torch.sqrt(torch.sum(torch.sum(p**2,1),1)))
        return x

class GroupSparseClassificationTrainer(ClassificationTrainer):
    def get_model(self, cfg=None, weights=None, verbose = True):
        model = GroupSparseClassificationModel(cfg, nc=self.data["nc"], verbose=verbose)
        return model
```

**Comment by glenn-jocher** [comment]
@xizzax hello! It looks like you're on the right track with customizing the YOLOv8 classification model. When saving your customized model, you should save the entire model object rather than just the state dictionary if you've made significant changes to the model architecture or loss function. This will ensure that all your customizations are preserved.

Here's how you can save and load the entire model:

```python
# Save the entire model
torch.save(new_model, 'my_custom_model.pth')

# Load the entire model
loaded_model = torch.load('my_custom_model.pth')
```

When you're ready to train, make sure to pass the path to your saved `.pth` file as the `model` argument:

```python
args = dict(model='my_custom_model.pth', data=dataset, epochs=1)
trainer = GroupSparseClassificationTrainer(overrides=args)
trainer.train()
```

This should allow you to train your customized model with your new loss function. If you encounter any issues during training, please provide the error messages, and we'll be happy to assist further. Good luck with your training! 🚀
